File: Solution1790.java
class Solution {
    public boolean areAlmostEqual(String s1, String s2) {
        if (s1 == null && s2 == null) return true;

        int total = 0;
        char[] result = new char[4];
        for (int i = 0; i < s1.length(); i++) {
            char c1 = s1.charAt(i);
            char c2 = s2.charAt(i);
            if (c1 == c2) continue;

            total++;
            if (total == 1) {
                result[0] = c1;
                result[1] = c2;
            } else if (total == 2) {
                result[2] = c1;
                result[3] = c2;
                if (result[0] != result[3] || result[1] != result[2]) return false;
            } else {
                return false;
            }
        }

        if (total == 0 || total == 2) return true;
        return false;
    }
}

File: rust/src/solution1790.rs
use crate::common::Solution;

impl Solution {
    pub fn are_almost_equal(s1: String, s2: String) -> bool {
        let mut total = 0;

        let mut result: [char; 4] = [' '; 4];

        let mut s1_array = s1.chars();
        let mut s2_array = s2.chars();
        for _ in 0..s1.len() {
            // trick -> nth will discard accessed value
            // e.g.  "abcde" -> nth(0) -> "bcde" -> nth(0) -> "cde"
            let c1 = s1_array.nth(0).expect("invalid index");
            let c2 = s2_array.nth(0).expect("invalid index");

            if c1 == c2 { continue; }

            total += 1;
            if total == 1 {
                result[0] = c1;
                result[1] = c2;
            } else if total == 2 {
                result[2] = c1;
                result[3] = c2;
            } else {
                return false
            }
        }

        if total == 0 { return true }
        if total == 2 && result[0] == result[3] && result[1] == result[2] { return true }
        false
    }
}